skimage/measure/_moments.py

```diff
@@ -240,10 +242,13 @@ def moments_central(image, center=None, order=3, **kwargs):
     """
     if center is None:
         center = centroid(image)
-    calc = image.astype(float)
+    float_dtype = _supported_float_type(image.dtype)
+    calc = image.astype(float_dtype, copy=False)
     for dim, dim_length in enumerate(image.shape):
-        delta = np.arange(dim_length, dtype=float) - center[dim]
-        powers_of_delta = delta[:, np.newaxis] ** np.arange(order + 1)
+        delta = np.arange(dim_length, dtype=float_dtype) - center[dim]
+        powers_of_delta = np.stack(
+            [delta ** c for c in range(order + 1)], axis=-1
+        )
```

Comment:
Why this change? :slightly_smiling_face:

Reply:
I was curious which was faster, but it is not very important to change this.

I timed it just now and it seems like new ways is faster for sizes larger than ~256, but the old way is faster for smaller sizes. Either way, I don't think this is a bottleneck, so I am fine with reverting this change if you prefer.


```Python
import numpy as np

def p1(delta, order=3):
    return delta[:, np.newaxis] ** np.arange(order + 1)

def p2(delta, order=3):
    return np.stack([delta[:, np.newaxis] ** o for o in np.arange(order + 1)], axis=-1)

for n in [10, 100, 256, 1000, 10000, 1000000]:
    delta = np.random.randn(n)
    print(f"\nSize {n}:")
    %timeit p1(delta)
    %timeit p2(delta)
```

```

Size 16:
3.07 µs ± 124 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
13.3 µs ± 448 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Size 64:
7.93 µs ± 293 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
15 µs ± 178 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Size 256:
18.4 µs ± 264 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
18.6 µs ± 424 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Size 1024:
65 µs ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
31.9 µs ± 816 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Size 4096:
250 µs ± 1.95 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
94.1 µs ± 2.88 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Size 16384:
1 ms ± 11.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
374 µs ± 8.29 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

Size 1048576:
68.7 ms ± 934 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
34.6 ms ± 725 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
the `order` parameter is usually small and previous implementation is one-line... Unless you think it is less readable, I suggest to keep previous version :wink:

Reply:
okay, the change has been reverted!